File: backend/s3.py
import os
import boto3
from botocore.exceptions import NoCredentialsError
from dotenv import load_dotenv

# Load environment variables from .env.local in the parent directory
dotenv_path = os.path.join(os.path.dirname(__file__), '..', '.env.local')
load_dotenv(dotenv_path=dotenv_path)

AWS_ACCESS_KEY_ID = os.getenv("AWS_ACCESS_KEY_ID")
AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
AWS_REGION = os.getenv("AWS_REGION", "us-east-1")
AWS_S3_BUCKET = os.getenv("AWS_S3_BUCKET")

if not all([AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_S3_BUCKET]):
    raise Exception("AWS credentials and/or S3 bucket name not found in environment variables")

s3_client = boto3.client(
    "s3",
    aws_access_key_id=AWS_ACCESS_KEY_ID,
    aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
    region_name=AWS_REGION
)

def download_from_s3(s3_key: str, local_path: str):
    """
    Download a file from S3
    """
    try:
        print(f"Downloading {s3_key} from S3 to {local_path}")
        s3_client.download_file(AWS_S3_BUCKET, s3_key, local_path)
        print("Download complete")
        return local_path
    except NoCredentialsError:
        print("Credentials not available")
        raise
    except Exception as e:
        print(f"Error downloading from S3: {e}")
        raise

def upload_file_to_s3(local_path: str, s3_key: str) -> str:
    """
    Upload a file to S3 and return its public URL
    """
    try:
        print(f"Uploading {local_path} to S3 with key {s3_key}")
        s3_client.upload_file(local_path, AWS_S3_BUCKET, s3_key)
        # Construct the URL
        url = f"https://{AWS_S3_BUCKET}.s3.{AWS_REGION}.amazonaws.com/{s3_key}"
        print(f"Upload complete: {url}")
        return url
    except NoCredentialsError:
        print("Credentials not available")
        raise
    except Exception as e:
        print(f"Error uploading to S3: {e}")
        raise

def generate_final_video_key(username: str, project_name: str) -> str:
    return f"users/{username}/{project_name}/final-video.mp4"

def get_s3_url(s3_key: str) -> str:
    """
    Get the public URL for an S3 object
    """
    return f"https://{AWS_S3_BUCKET}.s3.{AWS_REGION}.amazonaws.com/{s3_key}"

File: frontend/lib/s3.ts
import { PutObjectCommand, GetObjectCommand, S3Client } from "@aws-sdk/client-s3"
import { getSignedUrl } from "@aws-sdk/s3-request-presigner"
import fs from 'fs'
import { Readable } from 'stream'

function getS3Client() {
  const region = process.env.AWS_REGION
  const accessKeyId = process.env.AWS_ACCESS_KEY_ID
  const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY

  if (!region || !accessKeyId || !secretAccessKey) {
    throw new Error("Missing AWS env vars (AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY)")
  }

  return new S3Client({
    region,
    credentials: { accessKeyId, secretAccessKey },
  })
}

function getBucketName() {
  const bucket = process.env.S3_BUCKET_NAME
  if (!bucket) {
    throw new Error("Missing S3_BUCKET_NAME env var")
  }
  return bucket
}

export function generateS3Key(username: string, projectName: string, clipNumber: number) {
  const safeUser = username.replace(/[^a-zA-Z0-9-_]/g, "_")
  const safeProject = projectName.replace(/[^a-zA-Z0-9-_]/g, "_")
  return `${safeUser}/${safeProject}/clips/${clipNumber}.mp4`
}

export async function uploadToS3(buffer: Buffer, key: string, contentType: string) {
  const client = getS3Client()
  const bucket = getBucketName()

  await client.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: buffer,
      ContentType: contentType,
    })
  )

  return `s3://${bucket}/${key}`
}

export async function getPresignedUrl(key: string) {
  const client = getS3Client()
  const bucket = getBucketName()

  const command = new PutObjectCommand({
    Bucket: bucket,
    Key: key,
  })

  return getSignedUrl(client, command, { expiresIn: 3600 })
}

/**
 * Download a file from S3 to local path
 */
export async function downloadFromS3(s3Key: string, localPath: string): Promise<void> {
  const client = getS3Client()
  const bucket = getBucketName()

  const command = new GetObjectCommand({
    Bucket: bucket,
    Key: s3Key,
  })

  const response = await client.send(command)

  if (!response.Body) {
    throw new Error(`No data returned from S3 for key: ${s3Key}`)
  }

  // Convert body to stream and write to file
  const stream = response.Body as Readable
  const writeStream = fs.createWriteStream(localPath)

  return new Promise((resolve, reject) => {
    stream.pipe(writeStream)
    writeStream.on('finish', resolve)
    writeStream.on('error', reject)
  })
}

/**
 * Upload a file from local path to S3
 */
export async function uploadFileToS3(localPath: string, s3Key: string): Promise<string> {
  const client = getS3Client()
  const bucket = getBucketName()

  const fileContent = fs.readFileSync(localPath)

  await client.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: s3Key,
      Body: fileContent,
      ContentType: 'video/mp4',
    })
  )

  // Return public URL
  const region = process.env.AWS_REGION
  return `https://${bucket}.s3.${region}.amazonaws.com/${s3Key}`
}

/**
 * Generate S3 key for final video
 */
export function generateFinalVideoKey(username: string, projectName: string): string {
  const safeUser = username.replace(/[^a-zA-Z0-9-_]/g, "_")
  const safeProject = projectName.replace(/[^a-zA-Z0-9-_]/g, "_")
  return `${safeUser}/${safeProject}/final-video.mp4`
}

/**
 * Get public S3 URL for a key
 */
export function getS3Url(s3Key: string): string {
  const bucket = getBucketName()
  const region = process.env.AWS_REGION || 'us-east-1'
  return `https://${bucket}.s3.${region}.amazonaws.com/${s3Key}`
}
